### PR TITLE
Add Ruff formatter and pre-commit configurations

### PR DESCRIPTION
**Pull request by Zeitsperre**
# Changes

* Added a .pre-commit configuration
* Added configurations for `ruff`
* Ran safe and aggressive fixes over the codebase
* Add .idea folder to the `.gitignore`

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -16,4 +16,4 @@ jobs:
       - run: pip install poetry==1.1.11
       - run: poetry install
       - run: poetry run black --check .
```
Comment (bneijt) [inline .github/workflows/main.yml]: moving to ruff, we should remove black in favor of ruff-format.

Makes sense to me to just run pre-commit check here.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -16,4 +16,4 @@ jobs:
       - run: pip install poetry==1.1.11
```
Comment (bneijt) [inline .github/workflows/main.yml]: Really old version of poetry, probably why the pipeline PR pipeline is failing, let's bump this.

**Comment by bneijt** [comment]
Thank you for your great work and sorry for not getting back earlier. All changes seem sane to me, but pr pipeline is failing. Probably because it's on an old poetry version. I think if you fix those changes we could merge this asap.

**Comment by Zeitsperre** [comment]
@bneijt 

I've addressed your comments here. There are definitely some overlapping changes with #19, but I'll deal with those once this is merged. I've added a linting step to the `main.yml` so that if the checks don't pass linting, the rest of the tests won't run.

I've also updated the poetry version and regenerated the lock file with `--no-update` so that changes aren't as destructive.

The `ruff` formatter is reporting around 71 errors, mostly from bare exceptions and some undefined variables. I would suggest running the tests locally to ensure it looks good on your end, merging this Pull Request (overriding changes), then addressing the problems (or ignoring them) in #19.

All the best!

**Comment by Zeitsperre** [comment]
@bneijt 

I've added some `ignore` codes to the `ruff` configuration so that we can at the very least merge PR this without issue. The rest of the fixes are about ready in my other PR.

**Pull request merged**